### PR TITLE
[SYCL] Add test cases for SYCL2020 reductions

### PR DESCRIPTION
**Pull request by v-klochkov**
The corresponding changes in SYCL RT:
  intel/llvm#3410
  intel/llvm#3481

This patch
- adds checks for SYCL-2020 reductions with initialize_to_identity property (#3410)
- enables the test reduction_nd_N_vars.cpp for level-zero (#3481)
- reduces some of tests by eliminating unneeded/duplicated checks.
- removes the test reduction_transparent.cpp as useless because transparent 
  operators already checked in many other tests.
- reduces runtime of reduction tests by about 200x in average by re-using
  queue created once instead of creating it in each of test-cases

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Comment by v-klochkov** [comment]
This change is supposed to end annoying flaky reduction test fails in CI due to timeouts.

**Comment by rdeodhar** [comment]
The change to use a queue created in "main" looks good to me.

Regarding flakiness on the CPU device, could this message hold a clue for difference in behavior across machines?
"The implementation handling parallel_for with reduction requires work group size not bigger than 1"

File: SYCL/Reduction/reduction_nd_N_vars.cpp [hunk 1]
```diff
@@ -100,8 +97,9 @@ int testOne(T1 IdentityVal1, T1 InitVal1, BinaryOperation1 BOp1,
     CorrectOut2 = BOp2(CorrectOut2, InitVal2);
   if (Mode3 == access::mode::read_write)
     CorrectOut3 = BOp3(CorrectOut3, InitVal3);
-  // 4th reduction is USM and this is read_write.
-  CorrectOut4 = BOp4(CorrectOut4, InitVal4);
+  // discard_write mode for USM reductions is available only SYCL2020.
```
Comment (againull) [inline SYCL/Reduction/reduction_nd_N_vars.cpp:100]: I cannot match the comment to the code. I see USM but I don't see discard_write. Am I missing something?

**Comment by v-klochkov** [comment]
> The change to use a queue created in "main" looks good to me.
> 
> Regarding flakiness on the CPU device, could this message hold a clue for difference in behavior across machines?
> "The implementation handling parallel_for with reduction requires work group size not bigger than 1"

I see different (unexpected) output results. Combining that with diagnostics shown by you, it may be some problem with pointers, that I am going to debug/fix. That is not expected for ONEAPI::detail::reduGetMaxWGSize() to return 1 for those simple cases in reduction_nd_N_vars.cpp.

File: SYCL/Reduction/reduction_nd_s0_dw.cpp [hunk 2]
```diff
@@ -1,5 +1,4 @@
 // RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
-// RUNx: %HOST_RUN_PLACEHOLDER %t.out
```
Comment (againull) [inline SYCL/Reduction/reduction_nd_s0_dw.cpp:2]: Is this feature device-only? Or are we going to add support for HOST in the future? If the second, probably it makes sense to keep this line or add todo to not forget to enable it.

**Review by againull — COMMENTED** [review]
Also as a side note: it looks like there are too much different changes for one review. As far as I remember ideal PRs are < 300 lines. And in this case it even looks like merged changes from different PRs. I don't insist on splitting it, just a note.

File: SYCL/Reduction/reduction_nd_N_vars.cpp (hunk 1 above)
Comment (v-klochkov) [reply SYCL/Reduction/reduction_nd_N_vars.cpp:100]: ONEAPI::reduction if initialized with USM pointer assumed read-write to that USM memory. I.e. add the original value of element pointed by USM memory in the final sum. Only SYCL-2020 can ignore the original value of USM memory (i.e. do discard_write).

The test may have 'Mode4' be equal to access::mode::discard_write, which is treated as 'read_write' if 'IsSYCL2020==false'

File: SYCL/Reduction/reduction_nd_s0_dw.cpp (hunk 2 above)
Comment (v-klochkov) [reply SYCL/Reduction/reduction_nd_s0_dw.cpp:2]: running on host would require adding support for barrier() and group algorithm reduce(). That is a known issue that cannot be forgotten. Also several other tests already mention that (e.g. reduction_placeholder.cpp, reduction_queue_parallel_for.cpp )

Comment (v-klochkov) [reply SYCL/Reduction/reduction_nd_s0_dw.cpp:2]: I can return that line if you think it is useful to have it there.

**Comment by v-klochkov** [comment]
> Also as a side note: it looks like there are too much different changes for one review.

Yes, I know, sorry. That is not typical situation.
The LIT changes for https://github.com/intel/llvm/pull/3481 is removal of few lines disabling reduction_nd_N_vars.cpp for Windows and for level_zero on Linux.

The original change for https://github.com/intel/llvm/pull/3410 was smaller than that, but it did not pass several times without additional changes reducing tests and re-using queue. That is why it is so big now.

File: SYCL/Reduction/reduction_nd_N_vars.cpp (hunk 1 above)
Comment (v-klochkov) [reply SYCL/Reduction/reduction_nd_N_vars.cpp:100]: The comment could be this:
// ONEAPI::reduction supports only read_write access to USM memory

**Review by againull — APPROVED** [review]
Don't have any major concerns, LGTM

**Pull request merged**